### PR TITLE
For #2706: Refactor Glean to reduce errors

### PR DESCRIPTION
**Pull request by sblatz**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/fenix/blob/master/CHANGELOG.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt [hunk 1]
```diff
@@ -113,49 +121,55 @@ sealed class Event {
...
     }
 
     data class EnteredUrl(val autoCompleted: Boolean) : Event() {
-        override val extras: Map<String, String>?
-            get() = mapOf("autocomplete" to autoCompleted.toString())
+        override val extras: Map<Events.enteredUrlKeys, String>?
+            get() = mapOf(Events.enteredUrlKeys.autocomplete to autoCompleted.toString())
     }
 
+    // TODO: Do we want to change this mapping to strings? I think it *may* be necessary here.
```
Comment (sblatz) [inline app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt]: Would love your thoughts on this.

File: app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt [hunk 2]
```diff
@@ -245,16 +253,17 @@ sealed class Event {
...
     }
 
     sealed class Search
 
-    open val extras: Map<String, String>?
+    internal open val extras: Map<*, String>?
         get() = null
 }
 
+// TODO: Can we get rid of string mapping here?
```
Comment (sblatz) [inline app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt]: Same for string mapping here :)

File: app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt [hunk 3]
```diff
@@ -23,6 +23,7 @@ class FenixOnboarding(context: Context) {
 
     fun finish() {
         onboardingPrefs.onboardedVersion = CURRENT_ONBOARDING_VERSION
+        // TODO: This currently seems to be sent every time I test on debug...
```
Comment (sblatz) [inline app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt]: Any ideas what's going on here 🤔

File: app/src/main/java/org/mozilla/fenix/components/metrics/LeanplumMetricsService.kt [hunk 4]
```diff
@@ -29,8 +28,10 @@ private val Event.name: String?
         is Event.SyncAuthSignOut -> "E_Sign_Out_FxA"
         is Event.FXANewSignup -> "E_New_Sign_Up_FxA"
         is Event.ClearedPrivateData -> "E_Cleared_Private_Data"
+        is Event.DismissedOnboarding -> "E_Dismissed_Onboarding"
 
         // Do not track other events in Leanplum
+        // TODO: Do we want to explicitly list all Glean events? Or is this fine since it tends to be far less Leanplum than Glean?
```
Comment (sblatz) [inline app/src/main/java/org/mozilla/fenix/components/metrics/LeanplumMetricsService.kt]: Thoughts?

File: app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt [hunk 5]
```diff
@@ -304,8 +318,13 @@ private val Event.wrapper
             { Collections.renameButton.record(it) }
         )
 
-        // Don't track other events with Glean
-        else -> null
+        // Don't record other events in Glean:
+        is Event.AddBookmark -> { null }
```
Comment (boek) [inline app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt]: nit: I don't think we need brackets around null here

File: app/src/main/java/org/mozilla/fenix/components/metrics/LeanplumMetricsService.kt (hunk 4 above)
Comment (boek) [reply app/src/main/java/org/mozilla/fenix/components/metrics/LeanplumMetricsService.kt]: I think this is fine for Leanplum

File: app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt (hunk 1 above)
Comment (boek) [reply app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt]: These are for the values of extras :)

File: app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt (hunk 2 above)
Comment (boek) [reply app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt]: This is pattern matching the `Fact` which is from A-C converting it to an Event. So no need to change anything here. Tests wouldn't hurt though :D

File: app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt (hunk 3 above)
Comment (boek) [reply app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt]: Hmm, looks like this get's called even if onboarding isn't visible. We should probably fix in a follow up issue :)

**Comment by boek** [comment]
Can we get a data review for the new metrics (if there wasn't one already?) :)

File: app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt (hunk 3 above)
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt]: Filed follow up #4824

**Comment by sblatz** [comment]
Only new event we're adding is `selected` for search shortcuts. Here is the data review for it:

# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) **What questions will you answer with this data?**
- How often users select a search engine with the search shortcut menu

2) **Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?**

* To measure how often users select an engine from the search shortcut menu. This will help us determine if the feature is useful or should be tweaked in some way.

3) **What alternative methods did you consider to answer these questions? Why were they not sufficient?**

* N/A (These are baseline metrics)

4) **Can current instrumentation answer these questions?**
* Currently there are no events to track this event. We have events for opening and closing this menu but not for actually selecting an engine. 

5) **List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox [data c](https://wiki.mozilla.org/Firefox/Data_Collection)[ollection ](https://wiki.mozilla.org/Firefox/Data_Collection)[categories](https://wiki.mozilla.org/Firefox/Data_Collection) on the found on the Mozilla wiki.**

* All data is Category 2.

6) **How long will this data be collected?**

Until 03/01/2020

7) **What populations will you measure?**

* All release, beta, and nightly users with telemetry enabled.

8) **Please provide a general description of how you will analyze this data.**

* Glean / Amplitude

9) **Where do you intend to share the results of your analysis?**

* Only on Glean, Amplitude and with mobile teams.

File: app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt [hunk 6]
```diff
@@ -23,7 +22,9 @@ class FenixOnboarding(context: Context) {
 
     fun finish() {
         onboardingPrefs.onboardedVersion = CURRENT_ONBOARDING_VERSION
-        metrics.track(Event.DismissedOnboarding)
+
+        // To be fixed in #4824
```
Comment (sblatz) [inline app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt]: Intentionally leaving this here. Hoping I can pull this in this sprint.

**Pull request merged**